python/rmm/_lib/memory_resource.pyx

```diff
@@ -167,100 +163,50 @@ cdef class FixedSizeMemoryResource(MemoryResource):
         pass
 
 
-cdef class FixedMultiSizeMemoryResource(MemoryResource):
+cdef class BinningMemoryResource(MemoryResource):
     def __cinit__(
         self,
-        MemoryResource upstream,
-        size_t size_base=2,
-        size_t min_size_exponent=18,
-        size_t max_size_exponent=22,
-        size_t initial_blocks_per_size=128
+        MemoryResource upstream_mr
     ):
         self.c_obj.reset(
-            new thread_safe_resource_adaptor_wrapper(
-                shared_ptr[device_memory_resource_wrapper](
-                    new fixed_multisize_memory_resource_wrapper(
-                        upstream.c_obj,
-                        size_base,
-                        min_size_exponent,
-                        max_size_exponent,
-                        initial_blocks_per_size
-                    )
-                )
+            new binning_memory_resource_wrapper(
+                upstream_mr.c_obj
             )
         )
 
     def __init__(
         self,
-        MemoryResource upstream,
-        size_t size_base=2,
-        size_t min_size_exponent=18,
-        size_t max_size_exponent=22,
-        size_t initial_blocks_per_size=128
+        MemoryResource upstream_mr
     ):
-        """
-        Allocates blocks in the range `[min_size], max_size]` in power oftwo
-        steps, where `min_size` and `max_size` are both powers of two.
+        """"
+        Allocates memory from a set of specified "bin" sizes based on a specified allocation size.
+
+        Call add_bin to add fixed-size bin allocators.
 
         Parameters
         ----------
-        upstream : MemoryResource
-            The upstream memory resource used to allocate pools of blocks.
-        size_base : int, optional
-            The base of allocation block sizes (defaults is 2).
-        min_size_exponent : int, optional
-            The exponent of the minimum fixed block size to allocate
-            (default is 18).
-        max_size_exponent : int, optional
-            The exponent of the maximum fixed block size to allocate
-            (default is 22).
-        initial_blocks_per_size : int, optional
-            The number of blocks to preallocate from the upstream memory
-            resource, and to allocate when all current blocks are in use.
+        upstream_mr : MemoryResource
+            The memory resource to use for allocations larger than any of the bins
         """
         pass
 
-
-cdef class HybridMemoryResource(MemoryResource):
-    def __cinit__(
-        self,
-        MemoryResource small_alloc_mr,
-        MemoryResource large_alloc_mr,
-        size_t threshold_size=1<<22
+    cpdef add_bin(
```

Comment:
To expose the overload of `add_bin` that accepts a `device_memory_resource*`, something like the following would work:

```cython
cpdef add_bin(self, size_t allocation_size, object bin_resource=None):
    cdef MemoryResource _bin_resource

    if bin_resource is None:
        (<binning_memory_resource_wrapper*>(self.c_obj.get()))[0].add_bin(allocation_size)
    else:
        _bin_resource = bin_resource  # coerces Python object `bin_resource` to C object `_bin_resource`

        self.bin_resources.append(bin_resource)

        (<binning_memory_resource_wrapper*>(self.c_obj.get()))[0].add_bin(
            allocation_size,
            <device_memory_resource*>(_bin_resource.c_obj.get())
        )
```

Reply:
Thanks; done.